lib/codecs/src/encoding/format/csv.rs

```diff
@@ -0,0 +1,143 @@
+use crate::encoding::BuildError;
+use bytes::{BufMut, BytesMut};
+use lookup::lookup_v2::OwnedValuePath;
+use tokio_util::codec::Encoder;
+use vector_core::{
+    config::DataType,
+    event::{Event, Value},
+    schema,
+};
+
+/// Config used to build a `CsvSerializer`.
+#[crate::configurable_component]
+#[derive(Debug, Clone, Default)]
+pub struct CsvSerializerConfig {
+    /// The CSV fields.
+    pub fields: Vec<OwnedValuePath>,
+}
+
+impl CsvSerializerConfig {
+    /// Creates a new `CsvSerializerConfig`.
+    pub const fn new(fields: Vec<OwnedValuePath>) -> Self {
+        Self { fields }
+    }
+
+    /// Build the `CsvSerializer` from this configuration.
+    pub fn build(&self) -> Result<CsvSerializer, BuildError> {
+        if self.fields.is_empty() {
+            Err("At least one csv field must be specified".into())
+        } else {
+            Ok(CsvSerializer::new(self.fields.clone()))
+        }
+    }
+
+    /// The data type of events that are accepted by `JsonSerializer`.
+    pub fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement(&self) -> schema::Requirement {
+        // While technically we support `Value` variants that can't be losslessly serialized to
+        // CSV, we don't want to enforce that limitation to users yet.
+        schema::Requirement::empty()
+    }
+}
+
+/// Serializer that converts an `Event` to bytes using the CSV format.
+#[derive(Debug, Clone)]
+pub struct CsvSerializer {
+    fields: Vec<OwnedValuePath>,
+}
+
+impl CsvSerializer {
+    /// Creates a new `CsvSerializer`.
+    pub const fn new(fields: Vec<OwnedValuePath>) -> Self {
+        Self { fields }
+    }
+}
+
+impl Encoder<Event> for CsvSerializer {
+    type Error = vector_common::Error;
+
+    fn encode(&mut self, event: Event, buffer: &mut BytesMut) -> Result<(), Self::Error> {
+        let log = event.into_log();
+        let values = log.value();
+        let mut wtr = csv::Writer::from_writer(vec![]);
+
+        for field in &self.fields {
+            if let Some(value) = values.get(field) {
+                match value {
+                    Value::Bytes(bytes) => wtr.write_field(String::from_utf8(bytes.to_vec())?)?,
+                    Value::Integer(int) => wtr.write_field(int.to_string())?,
+                    Value::Float(float) => wtr.write_field(float.to_string())?,
+                    Value::Boolean(bool) => wtr.write_field(bool.to_string())?,
+                    Value::Timestamp(timestamp) => wtr.write_field(timestamp.to_rfc3339())?,
+                    Value::Null => wtr.write_field("NaN")?,
+                    // Array, Regex, Object are not supported by the CSV format.
+                    _ => wtr.write_field("NaN")?,
+                }
+            } else {
+                wtr.write_field("NaN")?;
+            }
```

Comment:
It may be more readable to `match` on the returned from `values.get(field)` - it took me a second to realize this was for `None`.